README.md
# MZQRCode
Swift二维码识别、生成、扫描、相册选择，使用简单，包含多码识别功能。

<div align=center>
<img src="1.gif" width="300px" />
</div>

#### Cocoapods 引入
```
pod 'MZQRCode', '~> 0.0.5'
```

#### 使用

- 扫码识别二维码
```
MZQRCodeManager.scanQRCode { value in
    print("扫码结果为:\(value)")
}
```

- 生成二维码
```
MZQRCodeManager.createQRCode(content: "zzz", size: 200.0, backgroundColor: .white, fillColor: .brown, logo: UIImage.init(named: "avatar")) { image in
    imageView.image = image
}
```

- 识别图片中二维码
```
MZQRCodeManager.identifyQRCode(imageView.image!) { content in
    print("识别结果为:\(content ?? "Null")")
}
```


CHANGELOG.md
### 0.0.5

- 更新时间

    2023年7月19日

- 新增

    1、增加国际化适配，包含中简、中繁、英语、日语

- 优化

    无


MZQRCode/ViewController.swift
//
//  ViewController.swift
//  MZQRCode
//
//  Created by 曾龙 on 2022/7/11.
//

import UIKit

class ViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let imageView = UIImageView.init(frame: CGRect(x: 0, y: 0, width: 300, height: 300))
        imageView.center = self.view.center
        imageView.contentMode = .scaleAspectFit
        imageView.isUserInteractionEnabled = true
        imageView.image = UIImage.init(named: "test")
        self.view.addSubview(imageView)
        
        let tap = UITapGestureRecognizer.init(target: self, action: #selector(identifyQRCode(tap:)))
        imageView.addGestureRecognizer(tap)
        
        let scanBtn = UIButton(type: .custom)
        scanBtn.frame = CGRect(x: 0, y: 120, width: 100, height: 40)
        var center = self.view.center
        center.y = scanBtn.center.y
        scanBtn.center = center
        scanBtn.setTitle("扫码识别", for: .normal)
        scanBtn.setTitleColor(.brown, for: .normal)
        scanBtn.addTarget(self, action: #selector(scanQRCode), for: .touchUpInside)
        self.view.addSubview(scanBtn)
        
        MZQRCodeManager.createQRCode(content: "zzz", fillColor: .brown, logo: UIImage.init(named: "avatar")) { image in
            imageView.image = image
        }
        
    }

    @objc func identifyQRCode(tap: UITapGestureRecognizer) {
        let imageView = tap.view as! UIImageView
        MZQRCodeManager.identifyQRCode(imageView.image!) { content in
            print("识别结果为:\(content ?? "Null")")
        }
    }
    
    @objc func scanQRCode() {
        MZQRCodeManager.scanQRCode { value in
            print("扫码结果为:\(value)")
        }
    }

}



MZQRCode/MZQRCode/Localized.swift
//
//  Localized.swift
//  MZQRCode
//
//  Created by macmini_zl on 2023/7/17.
//

import Foundation

extension String {
    
    func localized(
        _ language: Language = .current,
        value: String? = nil,
        table: String = "Localizable"
    ) -> String {
        guard let path = Bundle.qrBundle?.path(forResource: language.rawValue, ofType: "lproj") else {
            return self
        }
        return Bundle(path: path)?.localizedString(forKey: self, value: value, table: table) ?? self
    }
}

enum Language: String {
    case en
    case zhHans = "zh-Hans"
    case zhHant = "zh-Hant"
    case ja
    
    public static var current: Language = {
        guard let language = Locale.preferredLanguages.first else { return .en }
        
        if language.contains("ja") { return .ja }
        
        if language.contains("zh-HK") { return .zhHant }
        
        if language.contains("zh-Hant") { return .zhHant }
        
        if language.contains("zh-Hans") { return .zhHans }
        
        return Language(rawValue: language) ?? .en
    }()
}

extension Bundle {
    static let qrBundle: Bundle? = {
        let containnerBundle = Bundle(for: MZQRCodeScanController.self)
        return Bundle(path: containnerBundle.path(forResource: "MZQRCode", ofType: "bundle")!) ?? .main
    }()
}


MZQRCode/MZQRCode/MZQRCodeManager.swift
//
//  MZQRCodeManager.swift
//  MZQRCode
//
//  Created by 曾龙 on 2022/7/11.
//

import UIKit
import MZAuthorization

public struct MZQRCodeManager {
    
    /// 扫描二维码
    /// - Parameter completionHandler: 扫描回调
    public static func scanQRCode(completionHandler: @escaping (String) -> Void) {
        MZAuthorization.requestAuth(type: .camera) {
            let scanController = MZQRCodeScanController()
            scanController.completionHandler = completionHandler
            scanController.modalTransitionStyle = .coverVertical
            scanController.modalPresentationStyle = .fullScreen
            currentViewController()?.present(scanController, animated: true)
        }
    }
    
    /// 生成二维码图片
    /// - Parameters:
    ///   - content: 二维码内容
    ///   - size: 二维码尺寸
    ///   - logo: 二维码填充logo
    ///   - backgroundColor: 二维码背景颜色
    ///   - fillColor: 二维码颜色
    ///   - completionHandler: 二维码生成回调
    public static func createQRCode(content: String, size: CGFloat = 200.0, backgroundColor: UIColor = .white, fillColor: UIColor = .black, logo: UIImage? = nil, completionHandler: @escaping (UIImage) -> Void) {
        DispatchQueue.init(label: "createQRCode").async {
            // 创建滤镜对象
            let filter = CIFilter.init(name: "CIQRCodeGenerator")!
            // 恢复默认设置
            filter.setDefaults()
            // 设置数据
            let infoData = content.data(using: .utf8)
            filter.setValue(infoData, forKey: "inputMessage")
            // 生成二维码
            let outputImage = filter.outputImage
            
            // 设置二维码颜色
            let colorFilter = CIFilter.init(name: "CIFalseColor")!
            colorFilter.setDefaults()
            colorFilter.setValue(outputImage, forKey: "inputImage")
            colorFilter.setValue(CIColor(color: fillColor), forKey: "inputColor0")
            colorFilter.setValue(CIColor(color: backgroundColor), forKey: "inputColor1")
            
            let newOutputImage = colorFilter.outputImage!
            let scale = size / newOutputImage.extent.width
            
            var image = UIImage(ciImage: newOutputImage.transformed(by: CGAffineTransform(scaleX: scale, y: scale)))
            
            if let logoImage = logo {
                let rect = CGRect(x: 0, y: 0, width: image.size.width, height: image.size.height)
                
                UIGraphicsBeginImageContextWithOptions(rect.size, false, UIScreen.main.scale)
                image.draw(in: rect)
                let logoSize = CGSize(width: rect.width / 4, height: rect.height / 4)
                let x = (rect.width - logoSize.width) /  2
                let y = (rect.height - logoSize.height) / 2
                logoImage.draw(in: CGRect(x: x, y: y, width: logoSize.width, height: logoSize.height))
                
                image = UIGraphicsGetImageFromCurrentImageContext()!
                
                UIGraphicsEndImageContext()
            }
            
            DispatchQueue.main.async {
                completionHandler(image)
            }
        }
    }
    
    /// 识别二维码
    /// - Parameters:
    ///   - image: 二维码图片
    ///   - completionHandler: 识别回调
    public static func identifyQRCode(_ image: UIImage, completionHandler: @escaping (String?) -> Void) {
        DispatchQueue.init(label: "").async {
            var ciImage = image.ciImage
            if ciImage == nil {
                guard let cgImage = image.cgImage else {
                    DispatchQueue.main.async {
                        completionHandler(nil)
                    }
                    return
                }
                ciImage = CIImage.init(cgImage: cgImage)
            }
            
            // 初始化扫描仪，设置识别类型和识别质量
            let detector = CIDetector.init(ofType: CIDetectorTypeQRCode, context: nil, options: [CIDetectorAccuracy: CIDetectorAccuracyHigh])!
            
            // 扫描获取的特征组
            let features = detector.features(in: ciImage!)
            
            // 获取扫描结果
            if features.count == 1 {
                let feature = features[0] as! CIQRCodeFeature
                let content = feature.messageString
                DispatchQueue.main.async {
                    completionHandler(content)
                }
            } else if features.count > 1 {
                DispatchQueue.main.async {
                    let identifyController = MZQRCodeIndentifyController()
                    identifyController.completionHandler = completionHandler
                    identifyController.modalTransitionStyle = .coverVertical
                    identifyController.modalPresentationStyle = .fullScreen
                    identifyController.qrCodeImage = image
                    identifyController.features = features as? [CIQRCodeFeature]
                    currentViewController()?.present(identifyController, animated: true)
                }
            } else {
                DispatchQueue.main.async {
                    completionHandler(nil)
                }
            }
        }
    }
    
    private static func currentWindow() -> UIWindow? {
        let app = UIApplication.shared
        if let window = app.delegate?.window {
            return window
        } else {
            return app.keyWindow
        }
    }
    
    private static func currentViewController() -> UIViewController? {
        guard let controller = currentWindow()?.rootViewController else {
            return nil
        }
        return self.currentViewControllerFrom(controller)
    }
    
    private static func currentViewControllerFrom(_ root: UIViewController) -> UIViewController {
        let currentViewController: UIViewController
        if root.presentedViewController != nil {
            return self.currentViewControllerFrom(root.presentedViewController!)
        }
        if root.isKind(of: UITabBarController.classForCoder()) {
            currentViewController = self.currentViewControllerFrom((root as! UITabBarController).selectedViewController!)
        } else if root.isKind(of: UINavigationController.classForCoder()) {
            currentViewController = self.currentViewControllerFrom((root as! UINavigationController).visibleViewController!)
        } else {
            currentViewController = root
        }
        return currentViewController
    }
}


MZQRCode/MZQRCode/MZQRCodeIndentifyController.swift
//
//  MZQRCodeIndentifyController.swift
//  MZQRCode
//
//  Created by 曾龙 on 2022/7/12.
//

import UIKit

class MZQRCodeIndentifyController: UIViewController {
    
    private lazy var imageView: UIImageView = {
        let imageView = UIImageView(frame: self.view.bounds)
        imageView.contentMode = .scaleAspectFit
        return imageView
    }()
    
    var qrCodeImage: UIImage?
    var features: [CIQRCodeFeature]?
    var completionHandler: ((String) -> Void)?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = .black
        
        self.view.addSubview(self.imageView)
        self.imageView.image = qrCodeImage
        self.drawFeatures(features!)
    }
    
    func drawFeatures(_ features: [CIQRCodeFeature]) {
        
        let maskView = UIView.init(frame: self.view.bounds)
        maskView.backgroundColor = .gray.withAlphaComponent(0.5)
        self.view.addSubview(maskView)
        
        let cancelBtn = UIButton(type: .custom)
        cancelBtn.frame = CGRect(x: 10, y: MZ_STATUS_BAR_HEIGHT, width: 90, height: 50)
        cancelBtn.setTitle("Cancel".localized(), for: .normal)
        cancelBtn.addTarget(self, action: #selector(cancel(sender:)), for: .touchUpInside)
        cancelBtn.contentHorizontalAlignment = .left
        maskView.addSubview(cancelBtn)
        
        for feature in features {
            let button = UIButton(type: .custom)
            button.frame = self.makeFrame(feature: feature)
            button.setImage(UIImage.init(named: "arrow", in: .qrBundle, compatibleWith: nil), for: .normal)
            button.setTitle(feature.messageString, for: .reserved)
            button.addTarget(self, action: #selector(selectQRCode(sender:)), for: .touchUpInside)
            button.layer.add(self.scaleAnimation(), forKey: "basic")
            maskView.addSubview(button)
        }
        
        let tip = UILabel.init(frame: CGRect(x: 0, y: MZ_SCREEN_HEIGHT - MZ_SAFE_BOTTOM - 20 - 50, width: MZ_SCREEN_WIDTH, height: 50))
        tip.text = "Tip".localized()
        tip.textColor = .white
        tip.font = .systemFont(ofSize: 16)
        tip.textAlignment = .center
        tip.adjustsFontSizeToFitWidth = true
        tip.numberOfLines = 0
        maskView.addSubview(tip)
    }
    
    // 获取二维码在整个界面中的位置
    func makeFrame(feature: CIQRCodeFeature) -> CGRect {
        let startImageSize = self.getSize(self.qrCodeImage!)
        var fixelW: CGFloat = startImageSize.width
        var fixelH: CGFloat = startImageSize.height
        var topSpace = 0.0
        var leftSpace = 0.0
        let viewSize = self.view.bounds.size
        if fixelH / fixelW > viewSize.height / viewSize.width {
            fixelW = (fixelW / fixelH) * viewSize.height
            fixelH = viewSize.height
            leftSpace = (MZ_SCREEN_WIDTH - fixelW) / 2
        } else {
            fixelH = (fixelH / fixelW) * viewSize.width
            fixelW = viewSize.width
            topSpace = (MZ_SCREEN_HEIGHT - fixelH) / 2
        }
        let scale = fixelW / startImageSize.width
        
        var btnFrame = feature.bounds
        
        btnFrame = self.exchangeFrame(frame: btnFrame, image: self.qrCodeImage!)
        
        btnFrame = CGRect(x: btnFrame.origin.x * scale + leftSpace, y: btnFrame.origin.y * scale + topSpace, width: btnFrame.width * scale, height: btnFrame.height * scale)
        
        return btnFrame
    }
    
    // 获取图片尺寸
    func getSize(_ image: UIImage) -> CGSize {
        var fixelW = CGFloat(image.cgImage?.width ?? 0)
        var fixelH = CGFloat(image.cgImage?.height ?? 0)
        
        if image.imageOrientation == .left || image.imageOrientation == .right || image.imageOrientation == .leftMirrored || image.imageOrientation == .rightMirrored {
            fixelW = CGFloat(image.cgImage?.height ?? 0)
            fixelH = CGFloat(image.cgImage?.width ?? 0)
        }
        return CGSize(width: fixelW, height: fixelH)
    }
    
    // 获取二维码照片在图片中的位置
    func exchangeFrame(frame: CGRect, image: UIImage) -> CGRect {
        var newFrame = frame
        
        let fixelW: CGFloat = CGFloat(image.cgImage?.width ?? 0)
        let fixelH: CGFloat = CGFloat(image.cgImage?.height ?? 0)
        
        let x = frame.origin.x
        let y = frame.origin.y
        let w = frame.width
        let h = frame.height
        
        if image.imageOrientation == .up {
            newFrame.origin.y = fixelH - y - h
        } else if image.imageOrientation == .left {
            newFrame.origin.x = fixelH - y - h;
            newFrame.origin.y = fixelW - x - w;
        } else if image.imageOrientation == .right {
            newFrame.origin.x = y;
            newFrame.origin.y = x;
        } else if image.imageOrientation == .down {
            newFrame.origin.x = fixelW - x - w;
            newFrame.origin.y = y;
        } else if image.imageOrientation == .upMirrored {
            newFrame.origin.x = fixelW - x - w;
            newFrame.origin.y = fixelH - y - h;
        } else if image.imageOrientation == .downMirrored {
            newFrame.origin.y = y;
            newFrame.origin.x = x;
        } else if image.imageOrientation == .rightMirrored {
            newFrame.origin.x = fixelH - y - h;
            newFrame.origin.y = x;
        } else if image.imageOrientation == .leftMirrored {
            newFrame.origin.x = y;
            newFrame.origin.y = fixelW - x - w;
        }
        return newFrame
    }
    
    // 小绿点缩放动画
    func scaleAnimation() -> CABasicAnimation {
        let animation = CABasicAnimation.init(keyPath: "transform.scale")
        animation.fromValue = NSNumber.init(value: 1.0)
        animation.toValue = NSNumber.init(value: 0.7)
        animation.autoreverses = true
        animation.duration = 1.0
        animation.repeatCount = MAXFLOAT
        animation.isRemovedOnCompletion = false
        animation.fillMode = .forwards
        animation.timingFunction = CAMediaTimingFunction.init(name: CAMediaTimingFunctionName.easeIn)
        return animation
    }
    
    // 取消
    @objc func cancel(sender: UIButton) {
        self.dismiss(animated: true)
    }
    
    // 选择二维码
    @objc func selectQRCode(sender: UIButton) {
        let value = sender.title(for: .reserved)
        self.dismiss(animated: true) {
            self.completionHandler?(value ?? "")
        }
    }
    
}


MZQRCode/MZQRCode/MZQRCodeScanController.swift
//
//  MZQRCodeScanController.swift
//  MZQRCode
//
//  Created by 曾龙 on 2022/7/11.
//

import UIKit
import AVKit
import AVFoundation
import MZAuthorization
import PhotosUI

/// 屏幕宽度
let MZ_SCREEN_WIDTH = UIScreen.main.bounds.size.width

/// 屏幕高度
let MZ_SCREEN_HEIGHT = UIScreen.main.bounds.size.height

/// 状态栏高度
let MZ_STATUS_BAR_HEIGHT = UIApplication.shared.statusBarFrame.size.height

/// 底部安全区域高度
let MZ_SAFE_BOTTOM: CGFloat = {
    if #available(iOS 11.0, *) {
        return UIApplication.shared.keyWindow?.safeAreaInsets.bottom ?? 0
    }
    return 0
}()

class MZQRCodeScanController: UIViewController, AVCaptureMetadataOutputObjectsDelegate, PHPickerViewControllerDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    var completionHandler: ((String) -> Void)?
    
    private lazy var backButton: UIButton = {
        let button = UIButton(type: .custom)
        button.frame = CGRect(x: 0, y: MZ_STATUS_BAR_HEIGHT, width: 50, height: 50)
        button.setImage(UIImage(named: "back", in: .qrBundle, compatibleWith: nil), for: .normal)
        button.addTarget(self, action: #selector(back), for: .touchUpInside)
        return button
    }()
    
    private lazy var albumView: UIImageView = {
        let imageView = UIImageView(frame: CGRect(x: MZ_SCREEN_WIDTH - 20 - 64, y: MZ_SCREEN_HEIGHT - MZ_SAFE_BOTTOM - 20 - 64, width: 64, height: 64))
        imageView.image = UIImage(named: "album", in: .qrBundle, compatibleWith: nil)
        imageView.isUserInteractionEnabled = true
        let tap = UITapGestureRecognizer.init(target: self, action: #selector(showAlbum))
        imageView.addGestureRecognizer(tap)
        return imageView
    }()
    
    private lazy var line: UIImageView = {
        let imageView = UIImageView(frame: CGRect(x: 20, y: MZ_STATUS_BAR_HEIGHT + 90, width: MZ_SCREEN_WIDTH - 40, height: 20))
        imageView.contentMode = .scaleToFill
        imageView.image = UIImage.init(named: "scan", in: .qrBundle, compatibleWith: nil)
        return imageView
    }()
    
    private var device: AVCaptureDevice?
    private var input: AVCaptureDeviceInput?
    private var output: AVCaptureMetadataOutput?
    private var session: AVCaptureSession?
    private var preview: AVCaptureVideoPreviewLayer?
    private var isScaning: Bool = false
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = .black
        
        self.initPreview()
        
        self.view.addSubview(self.backButton)
        self.view.addSubview(self.albumView)
        self.view.addSubview(self.line)
    }
    
    func initPreview() {
        self.device = AVCaptureDevice.default(for: .video)
        try! self.input = AVCaptureDeviceInput.init(device: self.device!)
        self.output = AVCaptureMetadataOutput.init()
        self.output?.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
        self.session = AVCaptureSession.init()
        self.session?.sessionPreset = .hd1280x720
        if (self.session!.canAddInput(self.input!)) {
            self.session?.addInput(self.input!)
        }
        if (self.session!.canAddOutput(self.output!)) {
            self.session?.addOutput(self.output!)
        }
        self.output?.metadataObjectTypes = [.qr]
        self.output?.rectOfInterest = CGRect(x: 0, y: 0, width: 1, height: 1)
        
        self.preview = AVCaptureVideoPreviewLayer.init(session: self.session!)
        self.preview?.videoGravity = .resizeAspectFill
        self.preview?.frame = self.view.bounds
        self.view.layer.insertSublayer(self.preview!, at: 0)
    }
    
    //MARK: - AVCaptureMetadataOutputObjectsDelegate
    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        DispatchQueue.main.async {
            if !self.isScaning {
                return
            }
            self.stopScan()
            MZAudioPlayTool.vibrate()
            MZAudioPlayTool.playAudio("scan.wav")
            
            if metadataObjects.count == 1 {
                let result = metadataObjects[0] as! AVMetadataMachineReadableCodeObject
                self.dismiss(animated: true) {
                    self.completionHandler?(result.stringValue ?? "")
                }
            } else if (metadataObjects.count > 1) {
                var muchList = [Dictionary<String, Any>]()
                for item in metadataObjects {
                    let result = item as! AVMetadataMachineReadableCodeObject
                    var dic = Dictionary<String, Any>.init()
                    let code = result.stringValue
                    dic["code"] = code
                    
                    let frame = self.makeFrame(objc: result)
                    let frameStr = NSCoder.string(for: frame)
                    dic["frame"] = frameStr
                    muchList.append(dic)
                }
                self.drawMuchList(muchList)
            } else {
                self.startScan()
            }
        }
    }
    
    // 计算按钮位置
    func makeFrame(objc: AVMetadataMachineReadableCodeObject) -> CGRect {
        let isize = CGSize(width: 720.0, height: 1280.0)
        var Wout: CGFloat = 0
        var Hout: CGFloat = 0
        var wMore: Bool = true
        
        let previewSize = self.preview?.bounds.size ?? self.view.bounds.size
        
        if isize.width / isize.height > previewSize.width / previewSize.height {
            wMore = true
            Wout = (isize.width / isize.height) * previewSize.height
            Wout = Wout - previewSize.width
            Wout = Wout / 2
        } else {
            wMore = false
            Hout = (isize.height / isize.width) * previewSize.width
            Hout = Hout - previewSize.height
            Hout = Hout / 2
        }
        
        var point1: CGPoint = CGPoint.zero
        var point2: CGPoint = CGPoint.zero
        var point3: CGPoint = CGPoint.zero
        var point4: CGPoint = CGPoint.zero
        
        let array = objc.corners
        
        // 获取点
        for i in 0..<array.count {
            var point = array[i]
            
            point.x = point.y + point.x
            point.y = point.x - point.y
            point.x = point.x - point.y
            
            point.x = 1 - point.x
            
            if wMore {
                point.x = (point.x * (isize.width / isize.height) * previewSize.height) - Wout
                point.y = previewSize.height * point.y
            } else {
                point.x = previewSize.width * point.x
                point.y = (point.y * (isize.height / isize.width) * previewSize.width) - Hout
            }
            
            if i == 0 {
                point1 = point
            }
            if i == 1 {
                point2 = point
            }
            if i == 2 {
                point3 = point
            }
            if i == 3 {
                point4 = point
            }
        }
        
        let minX = min(point1.x, point2.x, point3.x, point4.x)
        let minY = min(point1.y, point2.y, point3.y, point4.y)
        let maxX = max(point1.x, point2.x, point3.x, point4.x)
        let maxY = max(point1.y, point2.y, point3.y, point4.y)
        
        let qrFrame = CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
        return qrFrame
    }
    
    func drawMuchList(_ muchList: [Dictionary<String, Any>]) {
        self.backButton.isHidden = true
        self.line.isHidden = true
        self.albumView.isHidden = true
        
        let maskView = UIView.init(frame: self.view.bounds)
        maskView.backgroundColor = .gray.withAlphaComponent(0.5)
        self.view.addSubview(maskView)
        
        let cancelBtn = UIButton(type: .custom)
        cancelBtn.frame = CGRect(x: 10, y: MZ_STATUS_BAR_HEIGHT, width: 90, height: 50)
        cancelBtn.setTitle("Cancel".localized(), for: .normal)
        cancelBtn.addTarget(self, action: #selector(cancel(sender:)), for: .touchUpInside)
        cancelBtn.contentHorizontalAlignment = .left
        maskView.addSubview(cancelBtn)
        
        for dic in muchList {
            let button = UIButton(type: .custom)
            button.frame = NSCoder.cgRect(for: dic["frame"] as! String)
            button.setImage(UIImage.init(named: "arrow", in: .qrBundle, compatibleWith: nil), for: .normal)
            button.setTitle(dic["code"] as? String, for: .reserved)
            button.addTarget(self, action: #selector(selectQRCode(sender:)), for: .touchUpInside)
            button.layer.add(self.scaleAnimation(), forKey: "basic")
            maskView.addSubview(button)
        }
        
        let tip = UILabel.init(frame: CGRect(x: 0, y: MZ_SCREEN_HEIGHT - MZ_SAFE_BOTTOM - 20 - 50, width: MZ_SCREEN_WIDTH, height: 50))
        tip.text = "Tip".localized()
        tip.textColor = .white
        tip.font = .systemFont(ofSize: 16)
        tip.textAlignment = .center
        tip.adjustsFontSizeToFitWidth = true
        tip.numberOfLines = 0
        maskView.addSubview(tip)
    }
    
    // 取消
    @objc func cancel(sender: UIButton) {
        sender.superview?.removeFromSuperview()
        self.backButton.isHidden = false
        self.line.isHidden = false
        self.albumView.isHidden = false
        self.startScan()
    }
    
    // 选择二维码
    @objc func selectQRCode(sender: UIButton) {
        let value = sender.title(for: .reserved)
        self.dismiss(animated: true) {
            self.completionHandler?(value ?? "")
        }
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.startScan()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        self.stopScan()
    }
    
    @objc func back() {
        self.dismiss(animated: true)
    }
    
    @objc func showAlbum() {
        MZAuthorization.requestAuth(type: .photoReadWrite) {
            if #available(iOS 14, *) {
                var configuration = PHPickerConfiguration()
                configuration.filter = .images
                configuration.selectionLimit = 1
                let picker = PHPickerViewController(configuration: configuration)
                picker.delegate = self
                self.present(picker, animated: true)
            } else {
                let picker = UIImagePickerController()
                picker.sourceType = .photoLibrary
                picker.delegate = self
                self.present(picker, animated: true)
            }
        }
    }
    
    func startScan() {
        self.isScaning = true
        self.preview?.session?.startRunning()
        self.startLineAnimation()
    }
    
    func stopScan() {
        self.isScaning = false
        self.preview?.session?.stopRunning()
        self.stopLineAnimation()
    }
    
    func startLineAnimation() {
        self.stopLineAnimation()
        
        let group = CAAnimationGroup.init()
        let scanAnimation = CABasicAnimation.init(keyPath: "position.y")
        scanAnimation.fromValue = NSNumber.init(value: MZ_STATUS_BAR_HEIGHT + 20)
        scanAnimation.toValue = NSNumber(value: MZ_SCREEN_HEIGHT - MZ_SAFE_BOTTOM - 20 - 64 - 20)
        scanAnimation.timingFunction = CAMediaTimingFunction.init(name: CAMediaTimingFunctionName.linear)
        
        let opacityAnimation = CABasicAnimation.init(keyPath: "opacity")
        opacityAnimation.fromValue = 1
        opacityAnimation.toValue = 0
        opacityAnimation.repeatCount = MAXFLOAT
        opacityAnimation.duration = 0.5
        opacityAnimation.beginTime = 2
        
        group.animations = [scanAnimation, opacityAnimation]
        group.isRemovedOnCompletion = false
        group.fillMode = .forwards
        group.duration = 2.5
        group.repeatCount = MAXFLOAT
        
        self.line.layer.add(group, forKey: "basic")
    }
    
    // 小绿点缩放动画
    func scaleAnimation() -> CABasicAnimation {
        let animation = CABasicAnimation.init(keyPath: "transform.scale")
        animation.fromValue = NSNumber.init(value: 1.0)
        animation.toValue = NSNumber.init(value: 0.7)
        animation.autoreverses = true
        animation.duration = 1.0
        animation.repeatCount = MAXFLOAT
        animation.isRemovedOnCompletion = false
        animation.fillMode = .forwards
        animation.timingFunction = CAMediaTimingFunction.init(name: CAMediaTimingFunctionName.easeIn)
        return animation
    }
    
    func stopLineAnimation() {
        self.line.layer.removeAllAnimations()
    }
    
    
    //MARK: - PHPickerViewControllerDelegate
    @available(iOS 14, *)
    func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
        picker.dismiss(animated: true) {
            for result in results {
                result.itemProvider.loadObject(ofClass: UIImage.self) { object, error in
                    if let image = object as? UIImage {
                        DispatchQueue.main.async {
                            MZQRCodeManager.identifyQRCode(image) { content in
                                if let message = content {
                                    self.dismiss(animated: true) {
                                        self.completionHandler?(message)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    
    //MARK: - UIImagePickerControllerDelegate
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picker.dismiss(animated: true) {
            if let image = info[.originalImage] as? UIImage {
                DispatchQueue.main.async {
                    MZQRCodeManager.identifyQRCode(image) { content in
                        if let message = content {
                            self.dismiss(animated: true) {
                                self.completionHandler?(message)
                            }
                        }
                    }
                }
            }
        }
    }
}
